irohad/ametsuchi/impl/storage_impl.cpp

```diff
@@ -93,30 +188,75 @@ namespace iroha {
           block_storage_factory_(std::move(block_storage_factory)),
           log_manager_(std::move(log_manager)),
           log_(log_manager_->getLogger()),
+          reconnection_strategy_factory_(
+              std::move(reconnection_strategy_factory)),
+          callback_factory_(std::make_unique<FailoverCallbackFactory>()),
           pool_size_(pool_size),
           prepared_blocks_enabled_(enable_prepared_blocks),
           block_is_prepared(false) {
       prepared_block_name_ =
           "prepared_block" + postgres_options_.dbname().value_or("");
 
-      for (size_t i = 0; i != pool_size_; i++) {
-        soci::session &session = connection_->at(i);
-        auto *backend = static_cast<soci::postgresql_session_backend *>(
-            session.get_backend());
-        PQsetNoticeProcessor(backend->conn_, &processPqNotice, log_.get());
-      }
+      /// function performs reconnection
+      auto connection_initialization_lambda =
+          [&](soci::session &session,
```

Comment:
```suggestion
          [log = log_](soci::session &session,
```
I was getting segmentation fault on access to `log_`, it can be captured by value instead.